invokeai/backend/model_manager/download/queue.py

```diff
@@ -0,0 +1,689 @@
+# Copyright (c) 2023, Lincoln D. Stein
+"""Implementation of multithreaded download queue for invokeai."""
+
+import os
+import re
+import shutil
+import threading
+import time
+import traceback
+from pathlib import Path
+from queue import PriorityQueue
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+import requests
+from huggingface_hub import HfApi, hf_hub_url
+from pydantic import Field, ValidationError, parse_obj_as, validator
+from pydantic.networks import AnyHttpUrl
+from requests import HTTPError
+
+from invokeai.app.services.config import InvokeAIAppConfig
+from invokeai.backend.util import InvokeAILogger, Logger
+
+from ..storage import DuplicateModelException
+from . import HTTP_RE, REPO_ID_RE
+from .base import (
+    DownloadEventHandler,
+    DownloadJobBase,
+    DownloadJobStatus,
+    DownloadQueueBase,
+    ModelSourceMetadata,
+    UnknownJobIDException,
+)
+
+# Maximum number of bytes to download during each call to requests.iter_content()
+DOWNLOAD_CHUNK_SIZE = 100000
+
+# marker that the queue is done and that thread should exit
+STOP_JOB = DownloadJobBase(id=-99, priority=-99, source="dummy", destination="/")
+
+# endpoint for civitai get-model API
+CIVITAI_MODEL_DOWNLOAD = r"https://civitai.com/api/download/models/(\d+)"
+CIVITAI_MODEL_PAGE = "https://civitai.com/models/"
+CIVITAI_MODEL_PAGE_WITH_VERSION = r"https://civitai.com/models/(\d+)\?modelVersionId=(\d+)"
+CIVITAI_MODELS_ENDPOINT = "https://civitai.com/api/v1/models/"
+CIVITAI_VERSIONS_ENDPOINT = "https://civitai.com/api/v1/model-versions/"
+
+
+class DownloadJobURL(DownloadJobBase):
+    """Job declaration for downloading individual URLs."""
+
+    source: AnyHttpUrl = Field(description="URL to download")
+
+
+class DownloadJobRepoID(DownloadJobBase):
+    """Download repo ids."""
+
+    variant: Optional[str] = Field(description="Variant, such as 'fp16', to download")
+    source: str = Field(description="URL or repo_id to download")
+
+    @validator("source")
+    @classmethod
+    def _validate_source(cls, v: str) -> str:
+        if not re.match(REPO_ID_RE, v):
+            raise ValidationError(f"{v} invalid repo_id", cls)
+        return v
+
+
+class DownloadJobPath(DownloadJobBase):
+    """Handle file paths."""
+
+    source: Union[str, Path] = Field(description="Path to a file or directory to install")
+
+
+class DownloadQueue(DownloadQueueBase):
+    """Class for queued download of models."""
+
+    _jobs: Dict[int, DownloadJobBase]
+    _worker_pool: Set[threading.Thread]
+    _queue: PriorityQueue
+    _lock: threading.RLock
```

Comment:
Added an explanation. It is because in some places there is an outer method that locks and then calls an inner method that tries to lock again. This avoids the inner method from blocking.

If I'm applying this pattern inappropriately I can try to find the places where there are two locks being applied and remove the inner one.